Add cluster topology documentation (#12)
ykakarap/kubecon-na-22-capi-lab: cluster-topology.md

> # Use Cluster API to update Cluster Topology

Reviewer: ```suggestion
# Use Cluster API to change the Cluster Topology
```
nit: Just have to have clearer separation between changing the topology vs update/upgrade Kubernetes

> [In the previous walk-through](your-first-cluster) we demonstrated how to create a cluster using a standard declarative configuration. Now we will show how to leverage this declarative configuration to update cluster topology with ease.

Reviewer: ```suggestion
[In the previous walk-through](./your-first-cluster.md) we demonstrated how to create a cluster using a standard declarative configuration. Now we will show how to leverage this declarative configuration to update cluster topology with ease.
```
nit

## What is Cluster Topology?

We define a cluster topology as the set of configurations that describe your cluster: a few examples are the number of control plane and worker nodes; the type of Machine hardware that underlies various nodes; regional distribution across nodes or node pools. You may also see this described as "cluster shape" elsewhere.

## Scale out control plane nodes

A common Kubernetes cluster maintenance activity is scaling out (or in) the number of control plane nodes in response to cluster activity. Because Cluster API configuration interfaces are in fact Kubernetes resources, there are a lot of ways to do this. We'll demonstrate using a variety of methods.

> ### Scale from 1 to 3 control plane nodes via kubectl apply

Reviewer: KCP doesn't allow 2 replicas because of quorum

> ### Scale from 1 to 3 control plane nodes via kubectl apply

Reviewer: We could go from 1 => 3 here and then back to 1 instead of up to 5 below though

Because we are leveraging the Kubernetes declarative model, we can simply refer to a desired configuration specification and rely upon Cluster API to evaluate what's different between the two.

Assuming that your `docker-cluster-one` is in its original configuration, you should have one control plane node:

```sh
$ kubectl --kubeconfig cluster-one.kubeconfig get nodes
NAME                                             STATUS   ROLES           AGE   VERSION
docker-cluster-one-md-0-nrh7k-75ddc4778f-vlph9   Ready    <none>          24m   v1.24.6
docker-cluster-one-xbqb2-tlmpb                   Ready    control-plane   24m   v1.24.6
```

Let's use the idempotent model to submit a modified configuration of our `docker-cluster-one` Cluster with 3 control plane replicas instead of 1. We've provided [a reference yaml of this updated configuration in this repo](yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml).

> If you're on macos or Linux you can diff this modified spec from the original spec used to create the `docker-cluster-one` Cluster:

Reviewer: ```suggestion
If you're on macOS or Linux you can diff this modified spec from the original spec used to create the `docker-cluster-one` Cluster:
```

```
$ diff yamls/clusters/docker-cluster-one.yaml yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml
17c17
<       replicas: 1
---
>       replicas: 3
```

> The above shows that the `yaml` spec are almost identical, with the only change being the `replicas` value on L17. By applying that modified spec to our kind management cluster we can achieve a control plane node scale out operation:

Reviewer: ```suggestion
The above shows that the `yaml` specs are almost identical, with the only change being the `replicas` value on L17. By applying that modified spec to our kind management cluster we can achieve a control plane node scale out operation:
```

docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          0s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          2s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          2s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          7s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          7s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          7s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready      <none>          10s   v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready      <none>          10s   v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready      <none>          12s   v1.24.6
```

### More scale operations

For such a simple cluster topology change against a single configuration, it's also possible to update our Cluster resource in-place. Let's use `kubectl edit` to do that.

```sh
$ kubectl --kubeconfig kubecon-na-22-capi-lab.kubeconfig edit cluster/docker-cluster-one
```

> The above command will engage your locally configured editor (for example, most Macos and Linux environments will be configured to launch `vim`). You'll want to look for the yaml configuration at the path `spec.topology.controlPlane.replicas`. Based on our prior scale out the value should be `3`. Go ahead and change that to `5`, and then save the changes in your editor:

Reviewer: ```suggestion
The above command will engage your locally configured editor (for example, most macOS and Linux environments will be configured to launch `vim`). You'll want to look for the yaml configuration at the path `spec.topology.controlPlane.replicas`. Based on our prior scale out the value should be `3`. Go ahead and change that to `5`, and then save the changes in your editor:
```

```sh
cluster.cluster.x-k8s.io/docker-cluster-one edited
```

After a few minutes we should now see 5 control plane nodes (note that we're pointing `kubectl` to our workload cluster below using the previously saved `cluster-one.kubeconfig` kubeconfig file):

```sh
$ kubectl --kubeconfig cluster-one.kubeconfig get nodes
NAME                                             STATUS   ROLES           AGE    VERSION
docker-cluster-one-md-0-nrh7k-75ddc4778f-vlph9   Ready    <none>          118m   v1.24.6
docker-cluster-one-xbqb2-8wcbj                   Ready    control-plane   31m    v1.24.6
docker-cluster-one-xbqb2-bcjvj                   Ready    control-plane   32s    v1.24.6
docker-cluster-one-xbqb2-fmwr8                   Ready    control-plane   20m    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready    control-plane   30m    v1.24.6
docker-cluster-one-xbqb2-xx9d8                   Ready    control-plane   102s   v1.24.6
```

> We can do to same gesture to scale out worker nodes. This time we want to edit the configuration at path `spec.topology.workers.machineDeployments`. We should only have one item in that array; change its `replicas` value from `1` to any other, larger value:

Reviewer: ```suggestion
We can do the same gesture to scale out worker nodes. This time we want to edit the configuration at path `spec.topology.workers.machineDeployments`. We should only have one item in that array; change its `replicas` value from `1` to any other, larger value:
```
nit

## What is Cluster Topology?

We define a cluster topology as the set of configurations that describe your cluster: a few examples are the number of control plane and worker nodes; the type of Machine hardware that underlies various nodes; regional distribution across nodes or node pools. You may also see this described as "cluster shape" elsewhere.

## Scale out control plane nodes

A common Kubernetes cluster maintenance activity is scaling out (or in) the number of control plane nodes in response to cluster activity. Because Cluster API configuration interfaces are in fact Kubernetes resources, there are a lot of ways to do this. We'll demonstrate using a variety of methods.

### Scale from 1 to 3 control plane nodes via kubectl apply

Because we are leveraging the Kubernetes declarative model, we can simply refer to a desired configuration specification and rely upon Cluster API to evaluate what's different between the two.

Assuming that your `docker-cluster-one` is in its original configuration, you should have one control plane node:

```sh

> $ kubectl --kubeconfig cluster-one.kubeconfig get nodes

Reviewer 2: Nit: how about breaking this into 2 sections - one for the command and one for the output. If we have a dedication one for the command then people will be able to use the "copy to clipboard button" in github and quickly copy and past the command the run the thing. 

Right now it copied the input and the output. People wont be able to use it unless they edit it locally. Would be nice to just copy paste and run.

Same comment at other places.

Author: Done

NAME                                             STATUS   ROLES           AGE   VERSION
docker-cluster-one-md-0-nrh7k-75ddc4778f-vlph9   Ready    <none>          24m   v1.24.6
docker-cluster-one-xbqb2-tlmpb                   Ready    control-plane   24m   v1.24.6
```

Let's use the idempotent model to submit a modified configuration of our `docker-cluster-one` Cluster with 3 control plane replicas instead of 1. We've provided [a reference yaml of this updated configuration in this repo](yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml).

> If you're on masOS or Linux you can diff this modified spec from the original spec used to create the `docker-cluster-one` Cluster:

Reviewer: I wonder if we should just simplify this sentence to something like. The diff between those two files is the following:

I wonder if it get's annoying if we have to repeat this sentence all over and I don't expect folks to run the diff command manually on top of the diff that we already provide

Author: Done

```
$ diff yamls/clusters/docker-cluster-one.yaml yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml
17c17
<       replicas: 1
---
>       replicas: 3
```

> The above shows that the `yaml` specs are almost identical, with the only change being the `replicas` value on L17. By applying that modified spec to our kind management cluster we can achieve a control plane node scale out operation:

Reviewer 2: ```suggestion
The above shows that the `yaml` specs are almost identical, with the only change being the `replicas` value on L17. By applying that modified spec to our kind management cluster we can achieve a control plane node scale up operation:
```
nit. Scale up is probably clearer. No strong preferences :) 
Same at other places.

Reviewer 2: ```suggestion
The above shows that the `yaml` specs are almost identical, with the only change being the `replicas` value on L17. By applying that modified spec to our kind management cluster we can achieve a control plane node scale up operation:
```
nit. Scale up is probably clearer. No strong preferences :) 

Same at other places.

Reviewer: in/out is horizontal scaling, up/down is vertical scaling

I would prefer the correct option :)

```sh
$ kind get kubeconfig --name kubecon-na-22-capi-lab > kubecon-na-22-capi-lab.kubeconfig
$ kubectl --kubeconfig kubecon-na-22-capi-lab.kubeconfig apply -f yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml
cluster.cluster.x-k8s.io/docker-cluster-one configured
```

Now we can watch the new control plane nodes come online:

```sh

> $ kubectl --kubeconfig cluster-one.kubeconfig get nodes -w

Reviewer 2: Looks like we are using `watch kubectl .... ` in other places. How about changing this also to use `watch` to be consistent?

Also, can we assume that `watch` is installed on windows? I think it is safe to assume that it will be available on linux and mac.

# Use Cluster API to change the Cluster Topology

[In the previous walk-through](./your-first-cluster.md) we demonstrated how to create a cluster using a standard declarative configuration. Now we will show how to leverage this declarative configuration to update cluster topology with ease.

**Table of Contents**

- [Scale out control plane nodes](#scale-out-control-plane-nodes)
  - [Scale from 1 to 3 control plane nodes](#scale-from-1-to-3-control-plane-nodes)
- [More scale operations](#more-scale-operations)
- [Add a new pool of worker nodes](#add-a-new-pool-of-worker-nodes)
- [Remove a worker node](#remove-a-worker-node)

> - Next: [Changing Cluster Topology](#changing-cluster-topology)

Reviewer: ```suggestion
- Next: [Changing Cluster Topology](./cluster-topology.md)
```

Author: lol this is a bug, should be the "MachineHealthChecks and Remediation" stuff.

Reviewer: We want nobody getting stuck in an infinite loop :D

kubectl --kubeconfig cluster-one.kubeconfig get nodes
```

Output:

```
NAME                                             STATUS   ROLES           AGE   VERSION
docker-cluster-one-md-0-nrh7k-75ddc4778f-vlph9   Ready    <none>          24m   v1.24.6
docker-cluster-one-xbqb2-tlmpb                   Ready    control-plane   24m   v1.24.6
```

Let's use the idempotent model to submit a modified configuration of our "`docker-cluster-one`" Cluster with 3 control plane replicas instead of 1. We've provided [a reference yaml of this updated configuration in this repo](yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml).

If you're on masOS or Linux you can diff this modified spec from the original spec used to create the "`docker-cluster-one`" Cluster:

```sh
diff yamls/clusters/docker-cluster-one.yaml yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml
```

Output:

```
17c17
<       replicas: 1
---
>       replicas: 3
```

The above shows that the `yaml` specs are almost identical, with the only change being the `replicas` value on L17. By applying that modified spec to our kind management cluster we can achieve a control plane node scale out operation:

```sh

> kind get kubeconfig --name kubecon-na-22-capi-lab > kubecon-na-22-capi-lab.kubeconfig
> kubectl --kubeconfig kubecon-na-22-capi-lab.kubeconfig apply -f yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml

Reviewer: ```suggestion
kubectl apply -f yamls/clusters/docker-cluster-one-3-control-plane-replicas.yaml
```

I think we can skip getting a kubeconfig for the mangement cluster 

After we created it, it is the current context. So we can keep it simple for users. kubectl current context is always the mgmt cluster and they then only need --kubeconfig for the workload clusters

Author: Ooh boy, you're really testing my OCD here....

Reviewer: I'll address this in a follow-up (assuming we wanted to, but otherwise we have to update a few more markdowns to always use the kubeconfig)

```

Output:

```
cluster.cluster.x-k8s.io/docker-cluster-one configured
```

Now we can watch the new control plane nodes come online:

```sh
kubectl --kubeconfig cluster-one.kubeconfig get nodes -w
```

> An interesting note! As your cluster transitions from 1 to 2 control plane nodes, you will temporarily lost etcd quorum and your the apiserver running on your cluster will briefly go offline. So your `kubectl -w` command will be interrupted. Re-run the same command again to resume watching your cluster nodes:

Reviewer: ```suggestion
An interesting note! As your cluster transitions from 1 to 2 control plane nodes, it will temporarily lose etcd quorum and the apiserver running on your cluster will briefly go offline. So your `kubectl -w` command will be interrupted. Re-run the same command again to resume watching your cluster nodes:
```
I think

Author: thx for the English lesson :)

docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          0s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          0s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          0s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          2s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          2s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          7s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          7s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   NotReady   <none>          7s    v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready      <none>          10s   v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready      <none>          10s   v1.24.6
docker-cluster-one-xbqb2-sjlwm                   Ready      <none>          12s   v1.24.6
```

## More scale operations

For such a simple cluster topology change against a single configuration, it's also possible to update our Cluster resource in-place. Let's use `kubectl edit` to do that.

```sh

> kubectl --kubeconfig kubecon-na-22-capi-lab.kubeconfig edit cluster/docker-cluster-one

Reviewer: As mentioned above. I would drop the `--kubeconfig kubecon-na-22-capi-lab.kubeconfig` here (and in the other cases below)

Reviewer: Adressed in follow-up

        infrastructure:
          ref:
            apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
            kind: DockerMachineTemplate
            name: quick-start-default-worker-machinetemplate
```

We'll reference the above "`default-worker`" class in [an updated Cluster configuration that declares a second worker pool](yamls/clusters/docker-cluster-one-second-worker-pool.yaml).

The difference between this spec and the original spec used to create the "`docker-cluster-one`" Cluster:

```sh
diff yamls/clusters/docker-cluster-one.yaml yamls/clusters/docker-cluster-one-second-worker-pool.yaml
```

Output:

```
35a36,38
>         - class: default-worker
>           name: md-1
>           replicas: 1
```

The above shows that we've added a new `MachineDeployment` called `md-1` (Our existing `MachineDeployment` is named `md-0`) to our new spec. By applying this modified spec to our kind management cluster we will initiate the creation of a new node, from this new node pool:

```sh
kubectl --kubeconfig kubecon-na-22-capi-lab.kubeconfig apply -f yamls/clusters/docker-cluster-one-second-worker-pool.yaml
```

Output:

```

> cluster.cluster.x-k8s.io/docker-cluster-one configured

Reviewer: Can't add a suggestion for that, but the closing ``` are missing

docker-cluster-one-md-1-ksqwt-8489684d5b-fpgjs   NotReady                   <none>          46s   v1.24.6
docker-cluster-one-md-1-ksqwt-8489684d5b-fpgjs   NotReady                   <none>          46s   v1.24.6
docker-cluster-one-md-1-ksqwt-8489684d5b-fpgjs   Ready                      <none>          76s   v1.24.6
```

## Remove a worker node

We can now demonstrate how easy it is to "roll back" such a change, as well as show how to remove an existing node pool from your Cluster configuration. In our case it's as easy as reapplying the original Cluster spec, which only declares one "`md-0`" node pool:

```sh
kubectl --kubeconfig kubecon-na-22-capi-lab.kubeconfig apply -f yamls/clusters/docker-cluster-one.yaml
```

Once again, we should observe only one running worker node, in the "`md-0`" pool:

```sh
kubectl --kubeconfig cluster-one.kubeconfig get nodes
```

Output:

```
NAME                                            STATUS   ROLES           AGE    VERSION
docker-cluster-one-md-0-lq4f8-b59497b9d-xchpm   Ready    <none>          99m    v1.24.6
docker-cluster-one-mvthd-k7dwf                  Ready    control-plane   174m   v1.24.6
```

## MachineHealthChecks and Remediation

> You are now in control of your Cluster's topology configuration! [Let's next explore MachineHealthChecks and Remediation to configure operational self-healing](machine-health-checks-remediation.md)...

Author: @sbueringer FYI, when we actually do the MachineHealthCheck stuff we'll want to circle back and make sure this link is valid (i.e., you might use a different name for the `.md` file...)

Reviewer: +1

I've added a sub-task that we should check the chaining in general
